src/coreclr/vm/methodtablebuilder.cpp

```diff
@@ -9898,6 +9897,21 @@ void MethodTableBuilder::CheckForSystemTypes()
 
                 return;
             }
+#if defined(UNIX_AMD64_ABI) || defined(TARGET_ARM64)
+            else if (strcmp(nameSpace, g_SystemNS) == 0)
+            {
+                EEClassLayoutInfo* pLayout = pClass->GetLayoutInfo();
+
+                // These types correspond to fundamental data types in the underlying ABIs:
+                // * Int128:  __int128
+                // * UInt128: unsigned __int128
+
+                if ((strcmp(name, g_Int128Name) == 0) || (strcmp(name, g_UInt128Name) == 0))
+                {
+                    pLayout->m_ManagedLargestAlignmentRequirementOfAllMembers = 16; // sizeof(__int128)
```

Comment:
This needs equivalent fix in crossgen/NativeAOT and in Mono

Reply:
@fanyang-mono could you point me to where `Mono` handles custom struct packing/alignment?

I tried checking for where that's handled for the vector types, but couldn't find it.

`Int128` and `UInt128` need to have 16-byte packing when targeting the `System V` ABI (used by Unix) and the same packing/alignment as `Int64`/`UInt64` otherwise (such as on Windows or 32-bit platforms).

Reply:
-- I fixed this up for crossgen/NativeAOT already and added corresponding P/Invoke tests to ensure the data is passed correctly to/from Native.

Reply:
It's handled in mono_class_layout_fields () in metadata/class-init.c.
Not sure the mono gc supports 16 byte aligned objects on 32 bit platforms.

Reply:
This doesn't impact 32-bit platforms, only 64-bit platforms (and only 64-bit Unix at that).

Reply:
@vargaz, looks like its just blocked everywhere not just 32-bit platforms: https://github.com/dotnet/runtime/blob/main/src/mono/mono/metadata/class-init.c#L2058-L2065

It looks like Mono also isn't differentiating "alignment" from "packing" and so the layout of types that include the new `Int128`, `UInt128`, or the existing `Vector64<T>`, `Vector128<T>`, `Vector256<T>` types will be incorrect.

This would explain why I couldn't find any logic touching `Vector128<T>` and ensuring that it has the right packing (and therefore layout) even if the GC couldn't correctly align the overall allocation.

Reply:
Yes, it looks like this is not implemented right now in mono.